nkms/storage/encrypted_file.py

```diff
@@ -0,0 +1,200 @@
+            encoded_header = msgpack.dumps(self.header)
+        except ValueError as e:
+            raise e
+        self.header_length = len(encoded_header)
+        return (self.header_length, encoded_header)
+
+    def _read_header(self):
+        """
+        Reads the header from the self.file_obj.
+        """
+        try:
+            # Read last four bytes (header length) of file.
+            self.file_obj.seek(-4, os.SEEK_END)
+
+            # The first four bytes of the file are the header length
+            self.header_length = int.from_bytes(
+                                     self.file_obj.read(4), byteorder='big')
+            # Seek to the beginning of the header and read it
+            self.file_obj.seek(-(self.header_length + 4), os.SEEK_END)
+            self.header = msgpack.loads(self.file_obj.read(self.header_length))
+        except ValueError as e:
+            self.file_obj.seek(0)
+            raise e
+        else:
+            # Seek to the end of the ciphertext
+            self.file_obj.seek(-(self.header_length + 4), os.SEEK_END)
+
+    def _read_chunk(self, chunk_size, nonce):
+        """
+        Reads a chunk and decrypts/authenticates it.
+
+        :param int chunk_size: Size of chunk to read from self.file_obj
+        :param bytes nonce: Nonce to use during decryption
+
+        :return: Decrypted/Authenticated chunk
+        :rtype: Bytes
+        """
+        ciphertext = self.file_obj.read(chunk_size)
+        return self.cipher.decrypt(ciphertext, nonce=nonce)
+
+    def open_new(self, keys, chunk_size=1000000, nonce=None):
+        """
+        Opens a new EncryptedFile and creates a header for it ready for
+        writing encrypted data.
+
+        :param list keys: Encrypted keys to put in the header.
+        :param int chunk_size: Size of encrypted chunks in bytes, default is 1MB
+        :param bytes nonce: 20 byte Nonce to use for encryption
+        """
+        self.file_obj = open(self.path, mode=self.mode)
+        self._build_header(nonce=nonce, keys=keys, chunk_size=chunk_size)
+
+    def open(self, is_new=False):
+        """
+        Opens a file for Encryption/Decryption.
+
+        :param bool is_new: Is the file new (and empty)?
+        """
+        # TODO: Error if self.file_obj is already defined
+        self.file_obj = open(self.path, mode=self.mode)
+
+        # file_obj is now ready for reading/writing encrypted data
+        if not is_new:
+            self._read_header()
+
+    def read(self, num_chunks=0):
+        """
+        Reads num_chunks of encrypted ciphertext and decrypt/authenticate it.
+
+        :param int num_chunks: Number of chunks to read. Default is all chunks
+
+        :return: List of decrypted/authenticated ciphertext chunks
+        :rtype: List
+        """
+        if num_chunks == 0:
+            num_chunks = self.header['chunks']
+
+        chunks = []
+        for chunk_num in range(num_chunks):
+            nonce = (self.header['nonce']
```

Comment:
I think that the next two lines can reasonably be turned into a method of their own.